rpc/client.go

```diff
@@ -91,219 +100,230 @@ type Client struct {
-			return err
-		}
+	go func() {
+		finishedCall := <-call.Done
+		newCall.ServiceMethod = finishedCall.ServiceMethod
+		newCall.Args = finishedCall.Args
+		newCall.Reply = finishedCall.Reply
+		newCall.Error = finishedCall.Error
 
-		conn, err := tlsDialHTTP(c.addr.Network(), c.addr.String(), tlsConfig)
-		if err != nil {
-			// Retry if the error is temporary, otherwise fail fast.
-			if t, ok := err.(net.Error); ok && t.Temporary() {
-				if log.V(1) {
-					log.Warning(err)
+		if newCall.Error != nil {
+			// From net/rpc/client.go; remote hung up. Our heartbeat runloop will
+			// reconnect, so this retryable.
+			if newCall.Error == io.ErrUnexpectedEOF {
+				newCall.Error = newRPCError(newCall.Error)
+			}
+		} else {
+			// Verify response data integrity if this is a proto response.
+			if resp, respOk := reply.(proto.Response); respOk {
+				if req, reqOk := args.(proto.Request); reqOk {
+					if err := resp.Verify(req); err != nil {
+						newCall.Error = err
+					}
 				}
-				continue
 			}
-			return err
 		}
+		newCall.Done <- newCall
+	}()
 
-		c.mu.Lock()
-		c.Client = rpc.NewClientWithCodec(codec.NewClientCodec(conn))
-		c.lAddr = conn.LocalAddr()
-		c.mu.Unlock()
-		if c.lAddr == nil {
-			return errClosed
-		}
-
-		// Ensure at least one heartbeat succeeds before exiting the
-		// retry loop. If it fails, don't retry: The node is probably
-		// dead.
-		if err = c.heartbeat(); err != nil {
-			return err
-		}
+	return newCall
+}
 
-		return nil
+func (c *Client) connect() error {
+	conn, err := tlsDialHTTP(c.network, c.remoteAddress, c.tlsConfig)
+	if err != nil {
+		return err
 	}
+	c.localAddress = conn.LocalAddr().String()
+	c.rpcClient = rpc.NewClientWithCodec(codec.NewClientCodec(conn))
 
-	return util.Errorf("system is stopping")
+	return nil
 }
 
-// IsConnected returns whether the client is connected.
-func (c *Client) IsConnected() bool {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.Client != nil
+// Healthy returns a channel that is closed when the client becomes healthy.
+func (c *Client) Healthy() <-chan struct{} {
+	return c.healthy.Load().(chan struct{})
 }
 
-// IsHealthy returns whether the client is healthy.
-func (c *Client) IsHealthy() bool {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.healthy
-}
-
-// Addr returns remote address of the client.
-func (c *Client) Addr() net.Addr {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.addr
-}
+func (c *Client) setHealthy() {
+	if c.isHealthy {
+		return
+	}
 
-// LocalAddr returns the local address of the client.
-func (c *Client) LocalAddr() net.Addr {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.lAddr
+	c.isHealthy = true
+	close(c.healthy.Load().(chan struct{}))
 }
 
-// RemoteOffset returns the most recently measured offset of the client clock
-// from the remote server clock.
-func (c *Client) RemoteOffset() proto.RemoteOffset {
-	c.mu.Lock()
-	defer c.mu.Unlock()
-	return c.offset
+func (c *Client) setUnhealthy() {
```

Comment:
`set{Unhealthy,Healthy}` can only ever be called from one place because they don't lock (and that place is the heartbeat loop) so maybe they should live in there for the time being.